portfolio/src/main/webapp/comments.js

```diff
@@ -0,0 +1,121 @@
+
+async function fetchComments() {
+  const response = await fetch('/comments');
+  const commentsJson = await response.json();
+  // create a hash table of type "comment id" -> "comment" to build a comment tree
+  const commentsMap = {};
+  for (const i in commentsJson) {
+    const comment = commentsJson[i];
+    commentsMap[comment.key] = comment;
+  }
+  if (commentsJson.length) {
+    // if there are some comments, remove label 'No comments yet'
+    document.getElementById('no-comments').remove();
+  }
+  // build a tree, starting DFS from the vertex (comments) that have no incoming vertexes (are not replies)
+  for (const i in commentsMap) {
+    const comment = commentsMap[i];
+    if (!comment.replyTo) {
+      buildCommentsTree(commentsMap, comment, 0);
+    }
+  }
+}
+
+function buildCommentsTree(commentsMap, comment, depth) {
+  // add current comment to dom
+  addCommentToDOM(comment, depth)
+  // iterate over all replies in the map, add them as the children of the current comment
+  for (const i in comment.replies) {
+    const replyId = comment.replies[i];
+    const reply = commentsMap[replyId]
+    buildCommentsTree(commentsMap, reply, Number(depth + 1));
+  }
+}
+
+function addCommentToDOM(comment, depth) {
```

Comment:
This function is very long. Can we split into multiple sub functions?

Reply:
made it a little bit shorter